docs/RingBuffer.md

```diff
@@ -0,0 +1,219 @@
+  - 32-2 = 30 bit record length limits records to 1GB.
+  - The lock and discard flag prevent the consumer from reading unfinished records.
+  - *Note:* This matches the linux record structure for future ring buffer mmap-consumer compatibility (see [#4163](https://github.com/microsoft/ebpf-for-windows/issues/4163)).
+- Records are padded to 8 byte alignment.
+  - 64 bit alignment is required for acquire/release semantics on 64 bit architectures.
+  - The consumer ignores and skips the padding (not included in the record length).
+
+### Producer functions
+
+Producers use reserve and submit to reserve space and submit completed records for reading.
+
+- The algorithms are described below in [Producer algorithm](#producer-algorithm).
+- `ebpf_ring_buffer_reserve` reserves a record and returns a pointer to the record data.
+  - Returns `EBPF_NO_MEMORY` if there isn't enough space left for the record.
+  - Wait-free with a single producer.
+    - This is the only potentially blocking producer/consumer function (retry wait at dispatch).
+- `ebpf_ring_buffer_submit` unlocks a record so that it can be read by the consumer
+- `ebpf_ring_buffer_discard` marks record as discarded and unlocks it so the consumer can skip it.
+- `ebpf_ring_buffer_output` calls reserve, copies the data into the record, then calls submit.
+  - *Note:* due to verifier limitations reserve/submit helper functions aren't available (only output, see #727 ).
+
+### Consumer functions
+
+The consumer can poll for new records to read using `ebpf_ring_buffer_next_consumer_record`,
+and return the memory for reuse after reading by passing the returned `next_offset` to `ebpf_ring_buffer_return_buffer`.
+
+- `ebpf_ring_buffer_next_consumer_record` also skips over any discarded records to return the memory to the ring.
+- *Note:* Currently only the `ring_buffer_output` helper is available, so there are never discarded records.
+
+A wait handle for producers to signal waiting consumers on is implemented at the eBPF map level,
+the internal ring buffer only handles the reading and writing of records.
+
+### Synchronization
+
+Specific memory ordering semantics are followed for record headers and the producer offset to
+ensure synchronization between producers and the consumer.
+
+- The consumer offset is only advanced by a single consumer, so doesn't have additional ordering constraints.
+
+1. Producers write-release the producer offset during reserve after locking the record.
+    - Write-release ensures the locked record header is visible before the producer offset is updated.
+2. Producers write-release the record header during submit/discard to unlock the record.
+    - Write-release ensures the record data is visible before the record is unlocked.
+    - This is also needed for discarded records to ensure any writes are completed before the space is reused.
+3. Consumer read-acquires the producer offset to check for available records.
```

Comment:
I notice that "producers" is plural in bullets 1 and 2, and "consumer" is singular in 3 and 4.  Does that mean this is designed for multiple producers and a single consumer?  In contrast line 114 and 134 use "consumers" plural.